go/vt/vtgate/engine/update.go

```diff
@@ -274,6 +215,59 @@ func (upd *Update) updateVindexEntries(vcursor VCursor, query string, bindVars m
 	return nil
 }
 
+// updateVindexEntries performs an update when a vindex is being modified
+// by the statement.
+// Note: the commit order may be different from the DML order because it's possible
+// for DMLs to reuse existing transactions.
+// Note 2: While changes are being committed, the changing row could be
+// unreachable by either the new or old column values.
+func (upd *Update) updateVindexEntriesScatter(vcursor VCursor, bindVars map[string]*querypb.BindVariable, rss []*srvtopo.ResolvedShard) error {
+	queries := make([]*querypb.BoundQuery, len(rss))
+	for i := range rss {
+		queries[i] = &querypb.BoundQuery{Sql: upd.OwnedVindexQuery, BindVariables: bindVars}
+	}
+	subQueryResult, errors := vcursor.ExecuteMultiShard(rss, queries, false, false)
+	for _, err := range errors {
+		if err != nil {
+			return vterrors.Wrap(err, "updateVindexEntriesScatter")
+		}
+	}
+
+	if len(subQueryResult.Rows) == 0 {
+		return nil
+	}
+
+	for _, row := range subQueryResult.Rows {
+		colnum := 1 // we start from the first non-vindex col
+		ksid := row[0].Raw()
+		for _, colVindex := range upd.Table.Owned {
+			// Fetch the column values. colnum must keep incrementing.
+			fromIds := make([]sqltypes.Value, 0, len(colVindex.Columns))
+			for range colVindex.Columns {
+				fromIds = append(fromIds, row[colnum])
+				colnum++
+			}
+
+			// Update columns only if they're being changed.
+			if colValues, ok := upd.ChangedVindexValues[colVindex.Name]; ok {
+				var vindexColumnKeys []sqltypes.Value
+				for _, colValue := range colValues {
+					resolvedVal, err := colValue.ResolveValue(bindVars)
+					if err != nil {
+						return err
+					}
+					vindexColumnKeys = append(vindexColumnKeys, resolvedVal)
+				}
+
+				if err := colVindex.Vindex.(vindexes.Lookup).Update(vcursor, fromIds, ksid, vindexColumnKeys); err != nil {
```

Comment:
It sucks that this is a single row API and `Delete` is multi-row. Let's change this to be multi-row, and pay the price of a breaking change. Otherwise, this will plague us forever.